cpp/src/arrow/types/struct.cc

```diff
@@ -17,4 +17,68 @@
 
 #include "arrow/types/struct.h"
 
-namespace arrow {}  // namespace arrow
+#include <sstream>
+
+namespace arrow {
+
+bool StructArray::EqualsExact(const StructArray& other) const {
+  if (this == &other) { return true; }
+  if (field_arrays_.size() != other.fields().size()) { return false; }
+  if (null_count_ != other.null_count_) { return false; }
+
+  if (null_count_ > 0) {
+    bool equal_null_bitmap =
+        null_bitmap_->Equals(*other.null_bitmap_, util::bytes_for_bits(length_));
+    if (!equal_null_bitmap) { return false; }
+
+    bool fields_equal = true;
+    for (size_t i = 0; i < field_arrays_.size(); ++i) {
+      fields_equal = field_arrays_[i].get()->Equals(other.field(i));
```

Comment:
This is actually somewhat problematic, since two arrays might be unequal but their unequal parts are "masked" by the parent bitmap. For example

```
struct A bitmap: 1 0 1 1
struct B bitmap: 1 0 1 1

struct A, child 0: 1 2 3 4
struct B, child 0, 1 5 3 4
```

so the data is technically equal, even though the children are different when you examine them without the "mask" of the parent bitmap

this suggests that the `Equals` method should accept an inclusion bitmap, which adds a lot of complexity. @emkornfield what do you think?